### PR TITLE
Expand and improve randomly-generated music descriptions

### PR DESCRIPTION
**Pull request by nornagon**
#### Summary
```SUMMARY: Content "Expand randomly-generated music descriptions"```

#### Purpose of change
The randomly-generated music descriptions were both inconsistent (generating descriptions with syntax errors like "neo -tune"), and were missing a period at the end of the description.

#### Describe the solution
I've extended the system of replacements in the snippets system so that all snippets can use the `<other_snippet_id>` syntax, not just NPC text, and then converted the music description logic to use it.

#### Describe alternatives you've considered
Perhaps this system isn't really necessary, and we'd rather stick to hand-authored lists of snippets rather than more complex generators such as this one.

The rules I've added to the JSON for describing music could be special-cased and written in C++ directly, rather than building a system for expanding text described in JSON.

Instead of reusing the snippets system for this sort of "expanding text", we could introduce a new JSON type ("snippet_generator"?) which could contain all the rules for generating a given type of snippet, like this music generator. This would prevent potential namespace overlap issues, since all the rules would be self-contained, but would prevent re-using generators in multiple places.

These more complex and more fragmented strings might present challenges for translators.

We could expand the system with additional capabilities, such as pluralization, article agreement (a/an), and state tracking, all of which Tracery implements. I've chosen to avoid those and write around these limitations for now to keep things simple for a first cut.

#### Additional context

Improved music descriptions:
Cataclysm__Dark_Days_Ahead_-_0_D-7603-g1b96ad13c5

### CODE REVIEW

**Comment by Qrox** [comment]
> These more complex and more fragmented strings might present challenges for translators

It would actually make translation relatively easier; previously three fragments were simply concatenated together, disregarding grammar variation among languages...

> #symbol#

We already have replacement tags in the form `<symbol>` in the snippet system. Please build the new system on top of that for better consistency. This also makes sure that the tags are specially handled on Transifex to avoid accidentally changing the tags during translation.

**Comment by nornagon** [comment]
It looks like currently the `<>` tags are only replaced when they're part of NPC talking or explicitly when generating map snippets for `<given_name>`, `<city>` and such, rather than a fully general system. But I agree that we shouldn't have two separate systems for this, so I'll look into consolidating them, using the `<>` syntax that already exists.

**Comment by nornagon** [comment]
There's still a bit of refactoring opportunity to have the npc talk tag replacement logic re-use the parsing from the snippet logic—e.g. passing a lambda to `snippet_library::expand` which will be called to handle unknown tags—but this works for now.

**Comment by ZhilkinSerg** [comment]
Some snippets-related tests are failing now.

File: src/text_snippets.cpp [hunk 1]
```diff
@@ -119,7 +141,7 @@ std::string snippet_library::random_from_category( const std::string &cat ) cons
     const int index = rng( 0, count - 1 );
     auto iter = iters.first;
     std::advance( iter, index );
-    return get( iter->second );
+    return expand( get( iter->second ) );
```
Comment (Qrox) [inline src/text_snippets.cpp]: It seems that the unit test wants this function to return the unexpanded snippet.

Comment (nornagon) [reply src/text_snippets.cpp]: I've updated the test to check something slightly different, and in the process I discovered:
1. it was testing two tags that were unused, `<drop_weapon>` and `<catch_up>`
2. there was a snippet using the nonexistent tag, `<catch_up>`.

I've removed the relevant snippet, and removed the old tags from the test.

As a follow-up, I'd like to write a test that iterates through every snippet and checks that any tags it references are either a special tag (like `<yrwp>`) or exist in the snippets library. That should prevent any undefined references from sneaking in.

File: data/json/npcs/talk_tags.json [hunk 2]
```diff
@@ -410,7 +410,6 @@
       "Stay with me, <name_g>!",
       "Catch up, <name_g>!",
       "Keep up!",
-      "Come on, <catch_up>!",
```
Comment (Qrox) [inline data/json/npcs/talk_tags.json]: Perhaps it was intended to use `<keep_up>` here.

File: tests/npc_test.cpp [hunk 3]
```diff
@@ -172,38 +172,29 @@ TEST_CASE( "on_load-similar-to-per-turn", "[.]" )
...
+        {
             "<name_b>", "<thirsty>", "<swear!>",
             "<sad>", "<greet>", "<no>",
             "<im_leaving_you>", "<ill_kill_you>", "<ill_die>",
             "<wait>", "<no_faction>", "<name_g>",
             "<keep_up>", "<yawn>", "<very>",
-            "<okay>", "<catch_up>", "<really>",
+            "<okay>", "<really>",
             "<let_me_pass>", "<done_mugging>", "<happy>",
-            "<drop_weapon>", "<swear>", "<lets_talk>",
```
Comment (Qrox) [inline tests/npc_test.cpp]: `<drop_it>` was probably the intended value.

File: data/json/npcs/talk_tags.json (hunk 2 above)
Comment (nornagon) [reply data/json/npcs/talk_tags.json]: Yep, but I decided to delete it instead because the capitalization would be wrong, e.g. `Come on, Keep up!!`

**Comment by nornagon** [comment]
I've updated the PR description to reflect the new code, and I've addressed all review comments. Please take another look when you can :)

**Pull request merged**